Homework assignment for checkboxes.

pavlo-kuvshynkin opened the pull request:
Completed automating all 3 test cases for the checkboxes homework assignment. Please, let me know if there are places to improve or refactor the code to enhance scripting. Thanks!

pavlo-kuvshynkin committed:
Homework assignment for checkboxes.

ba-instructor commented on tests/checkBoxes.spec.ts:
```diff
@@ -0,0 +1,78 @@
+
+    test('Test Case 1: Validate selected specialties', async ({page}) => {
+        //3. Locating targeted row and clicking 'Edit'
+        const tableRows = page.getByRole('row', { name: 'Helen Leary' })
```
It's not a `tableRows`. It's a signle row for Helen Leary. Also this constant is not needed at all since it was used just once in the test in the next line. So you could just write a single like to select that Edit Vet button

ba-instructor commented on tests/checkBoxes.spec.ts:
```diff
@@ -0,0 +1,78 @@
+        const tableRows = page.getByRole('row', { name: 'Helen Leary' })
+        await tableRows.getByRole('button', {name: "Edit Vet"}).click()
+        //4. Asserting specialties fields has value 'radiology'
+        const specialtiesField = page.locator('.dropdown .selected-specialties')
```
Locator can be simplified to just `.selected-specialties`

ba-instructor commented on tests/checkBoxes.spec.ts:
```diff
@@ -0,0 +1,78 @@
+        //5. Clicking specialties field to show up menu with checkboxes
+        await specialtiesField.click()
+        //6. Asserting 'radiology' is checked
+        const checkBox1 = page.getByRole('checkbox', {name:'radiology'})
```
This is a very bad naming pattern like 1, 2, 3 and so on. Always provide a good meaningful naming for the constants, variables, and method names.

ba-instructor commented on tests/checkBoxes.spec.ts:
```diff
@@ -0,0 +1,78 @@
+    test('Test Case 2: Select all specialties', async ({page}) => {
+        //2. Locating targeted row and clicking 'Edit'
+        const tableRows = page.getByRole('row', { name: 'Rafael Ortega' })
+        await tableRows.getByRole('button', {name: "Edit Vet"}).click()
```
Same here about naming and constant reusability

ba-instructor commented on tests/checkBoxes.spec.ts:
```diff
@@ -0,0 +1,78 @@
+            expect(await box.isChecked()).toBeFalsy()
+        }
+        //6. Asserting there are no specialties filed
+        await expect(specialtiesField).toHaveText('')
```
Check playwright documentation and find a more suitable assertion to validate that there is no text in the input field.

ba-instructor reviewed (commented):
Please check the comments.

I didn't put notes to all occurrences in the code that should be fixed. Review the rest of the code, and make adjustments accordingly.

Main feedback - don't make shortcuts on naming for your constants, variables, and method names. If you get used to bad naming, it will hit you hard later when you write a more complex code, and then later will be struggling to understand what is it doing because of the bad naming.

pavlo-kuvshynkin committed:
Changes according to the comments on pointed mistakes.

pavlo-kuvshynkin commented:
Applied changes in both files to the consts naming to be more readable. Also, removed removed locator consts with one time usage and added better validation for empty input fields.

ba-instructor commented on tests/inputFields.spec.ts:
```diff
@@ -9,8 +9,8 @@ test.describe('Input fields', () => {
         //Creating locator for targeted table row that is going to be edited
-        const tableRows = page.locator('tbody tr')
-        await tableRows.getByRole('button',{name: "Edit"}).first().click()
+        const tableRow = page.locator('tbody tr')
```
In this example, the initial naming was correct, but you have changed it to incorrect :)
This locator returns the list of table rows, instead of a single table row, so the "`tableRows`" was more correct constant name.

ba-instructor commented on tests/inputFields.spec.ts:
```diff
@@ -22,22 +22,22 @@ test.describe('Input fields', () => {
-        const tableRows = page.getByRole('row', { name: 'dog' })
-        await tableRows.getByRole('button', {name: "Edit"}).click()
+        //Creating locator for targeted table row that is going to be edited
+        const tableRow = page.getByRole('row', { name: 'dog' })
```
When picking the names, try to be as specific as possible, it will not hurt.
So this locator is not just some table row. It is a `dogTableRow` or `tableRowOfDogType`, something like that.

It may seem like a not big deal to you right now, but trust me, descriptive naming helps a lot later, when you read more complex code which you haven't see for one month and forgot what was it about.

pavlo-kuvshynkin replied:
I thought for the items that contain dynamic values it is not a good idea to add any value to the name of the locator as it may be changed at some point. Maybe a better approach is to make it a more specific name like `petTypeTableRow`?

ba-instructor replied:
It's no longer a dynamic value. This constant points to the row with a "dog." It can't resolve anything else, "cat" row for example. If the "dog" value is not there, the locator will just fail.

ba-instructor reviewed (approved):
Looks good, I approve this, but check a couple of comments related to the naming one more time. Make adjustments and merge it.

pavlo-kuvshynkin committed:
Added adjusted locator names to make it more specific.

pavlo-kuvshynkin committed:
TR 'dog'
